client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java

```diff
@@ -247,23 +126,40 @@ protected RestHighLevelClient(RestClientBuilder restClientBuilder, List<NamedXCo
      */
     protected RestHighLevelClient(RestClient restClient, CheckedConsumer<RestClient, IOException> doClose,
                                   List<NamedXContentRegistry.Entry> namedXContentEntries) {
-        this.client = Objects.requireNonNull(restClient, "restClient must not be null");
-        this.doClose = Objects.requireNonNull(doClose, "doClose consumer must not be null");
-        this.registry = new NamedXContentRegistry(
-                Stream.of(getDefaultNamedXContents().stream(), getProvidedNamedXContents().stream(), namedXContentEntries.stream())
-                    .flatMap(Function.identity()).collect(toList()));
+        requestActions = new RestRequestActions(restClient, doClose, namedXContentEntries);
+        indicesClient = new IndicesClient(requestActions);
+        clusterClient = new ClusterClient(requestActions);
+        ingestClient = new IngestClient(requestActions);
+        snapshotClient = new SnapshotClient(requestActions);
+        tasksClient = new TasksClient(requestActions);
+        xPackClient = new XPackClient(requestActions);
+        watcherClient = new WatcherClient(requestActions);
+        graphClient = new GraphClient(requestActions);
+        licenseClient = new LicenseClient(requestActions);
+        migrationClient = new MigrationClient(requestActions);
+        machineLearningClient = new MachineLearningClient(requestActions);
+        securityClient = new SecurityClient(requestActions);
+        rollupClient = new RollupClient(requestActions);
     }
 
     /**
      * Returns the low-level client that the current high-level client instance is using to perform requests
      */
     public final RestClient getLowLevelClient() {
-        return client;
+        return requestActions.getClient();
+    }
+
+    /**
+     * Returns the wrapper used to communicate with the low-level client that all subclients should be using.
+     */
+    public final RestRequestActions getRequestActions() {
```

Comment:
I'd be more comfortable if this were protected or a public method like `<T> T buildExtension(Function<RestRequestActions, T>)`.

Reply:
In 054deea I dropped the visibility of `getRequestActions`, and added the `buildSubClient` method like you mention above.